docs/design/data_safety.md

```diff
@@ -0,0 +1,153 @@
+1. _Whereas_ algorithms that panic on malformed data increase the vulnerability space of an application\*\*; and
+1. _Whereas_ "garbage in, garbage out" (GIGO) does not increase the space that malicious actors could leverage\*\*; be it
+1. _Resolved_ that code should never panic at runtime based on invalid data; and be it
+1. _Resolved_ that data structs should reduce the number of internal invariants; and be it
+1. _Resolved_ that code paths only reachable by invalid data should use GIGO with debug assertions.
+
+\* *As a thought experiment, if you were 100% confident, you could use `get_unchecked` and other unsafe operations. If you are not confident enough to use unsafe code, then you are not 100% confident.*
+
+\*\* *An attacker could construct data to produce a result they desire, regardless of whether unexpected operations end in GIGO. However, panicky algorithms could enable attackers to perform denial-of-service attacks.*
+
+Additional points:
+
+1. If performing binary search on a vector from data, it is not necessary to validate that the vector is sorted. The binary search will fail to find elements in an unsorted vector, but it won't panic, so this is consistent with GIGO.
+2. The best data structs are those that don't need to be validated. See the below example on how to write data structs without internal invariants.
+
+## Example
+
+Consider the struct
+
+```rust
+#[derive(serde::Deserialize)]
+pub struct WeekdayNamesBad {
+    // Invariant 1: the first element in the vector is the default value
+    // Invariant 2: there are 7 elements in the vector
+    weekday_names: Vec<String>,
+}
+
+impl WeekdayNamesBad {
+    pub fn get_first_weekday_name(&self) -> &str {
+        &self.weekday_names[0]
+    }
+    pub fn get_weekday_name_at_index(&self, idx: usize) -> Option<&str> {
+        if idx >= 7 {
+            return None;
+        }
+        Some(&self.weekday_names[idx])
+    }
+}
+```
+
+This struct has 2 invariants, and each of the functions could panic if those invariants are not upheld.
+
+Below are 3 approaches to resolve this issue that are consistent with the ICU4X style.
+
+*Editor's note: the examples in this document are tested along with the tutorials.*
+
+### Solution 1: Garbage In, Garbage Out
+
+Change the functions to return default fallback values if the data is not in the expected form.
+
+```rust
+#[derive(serde::Deserialize)]
+pub struct WeekdayNamesGIGO {
+    // WEAK Invariant 1: the first element in the vector is the default value
+    // WEAK Invariant 2: there are 7 elements in the vector
+    weekday_names: Vec<String>,
+}
+
+impl WeekdayNamesGIGO {
+    pub fn get_first_weekday_name(&self) -> &str {
+        match self.weekday_names.get(0) {
+            Some(v) => v,
+            None => {
+                debug_assert!(false, "weekday_names is empty");
+                ""
+            }
+        }
+    }
+    pub fn get_weekday_name_at_index(&self, idx: usize) -> Option<&str> {
```

Comment:
For a good example, this method should document that the client can not assume that `idx` less than 7 can be unwrapped. This is a quick assumption to make when dealing with weekdays, however if the client treats this like the non-gigo data struct and does something like `get_weekday_name_at_index(4).expect("week has 7 days")`, you've just moved the panic.

Reply:
I added a docs comment to the function in the bad version.